src/java/org/apache/cassandra/repair/Scheduler.java

```diff
@@ -0,0 +1,181 @@
+            long smallestScore = -1;
+            for (var g : groups.values())
+            {
+                if (g.isEmpty())
+                    continue;
+                if (smallest == null)
+                {
+                    smallest = g;
+                    smallestScore = g.score();
+                }
+                else
+                {
+                    var score = g.score();
+                    if (score < smallestScore)
+                    {
+                        smallest = g;
+                        smallestScore = score;
+                    }
+                }
+            }
+            if (smallest == null)
+                return;
+            inflight++;
+            smallest.executeNext();
+        }
+
+        private class Group
+        {
+            private final TimeUUID sessionId;
+            private final Executor taskExecutor;
+            private final List<Task<?>> tasks = new ArrayList<>();
+            private final LongArrayList durations = new LongArrayList();
+            private int inflight = 0;
+            private int completed = 0;
+
+            private Group(TimeUUID sessionId, Executor taskExecutor)
+            {
+                this.sessionId = sessionId;
+                this.taskExecutor = taskExecutor;
+            }
+
+            public long score()
+            {
+                if (tasks.isEmpty())
+                    return -1;
+                long avgDuration = (long) durations.longStream().average().orElse(TimeUnit.HOURS.toNanos(1));
+                return tasks.size() * avgDuration;
+            }
+
+            public void executeNext()
+            {
+                Task<?> task = tasks.get(0);
+                tasks.remove(0);
+                inflight++;
+                var startNs = ctx.clock().nanoTime();
+                task.addCallback((s, f) -> onDone(this, ctx.clock().nanoTime() - startNs));
+                taskExecutor.execute(task);
+            }
+
+            public void add(Task<?> task)
+            {
+                tasks.add(task);
+            }
+
+            private void update(long durationNs)
+            {
+                durations.add(durationNs);
+                inflight--;
+                completed++;
+            }
+
+            public boolean isEmpty()
+            {
+                return tasks.isEmpty();
+            }
+
+            @Override
+            public String toString()
+            {
+                return "Group{" +
+                       "sessionId=" + sessionId +
+                       ", tasks=" + tasks.size() +
+                       ", durations=" + durations.longStream().average().orElse(-1) +
+                       ", score=" + score() +
+                       ", inflight=" + inflight +
+                       ", completed=" + completed +
+                       '}';
+            }
+        }
+    }
+
+    abstract class Task<T> extends AsyncFuture<T> implements Runnable
```

Comment:
isn't this just `org.apache.cassandra.concurrent.FutureTask`?

Reply:
This is slightly different due to how the `run` method sets a callback. I have moved the setting of that callback into the `Task` class to encapsulate it better.